# Pandas
Pandas is a high performance easy to use tool for data analysis in python programming.
Pandas provides two important data structures:
1. Series: A pandas series is a single column (Similar to lists).
2. Dataframes: Can be considered as relational table with rows representing an instance of an example and columns representing different variables

In [1]:
from __future__ import print_function
import pandas as pd
import matplotlib.pyplot as plt 
import numpy as np

## Series
A pandas series is a single column. Similar to lists

In [2]:
city_names = pd.Series(['New Delhi', 'Mumbai', 'New York', 'London', 'Tokyo', 'Seoul', 'San francisco', 'Sydney', 'Dubai'])
city_names

0        New Delhi
1           Mumbai
2         New York
3           London
4            Tokyo
5            Seoul
6    San francisco
7           Sydney
8            Dubai
dtype: object

In [3]:
population_in_M = pd.Series([18.7, 23.6, 22, 8.7, 13.5, 10.3, 8.75, 5.2, 3.38])
population_in_M

0    18.70
1    23.60
2    22.00
3     8.70
4    13.50
5    10.30
6     8.75
7     5.20
8     3.38
dtype: float64

## Dataframe
Can be considered as relational table with rows representing an instance of an example and columns representing different variables

In [4]:
cities = pd.DataFrame({'City_names': city_names, 'Population_in_M' : population_in_M})
cities

City_names  Population_in_M
0      New Delhi            18.70
1         Mumbai            23.60
2       New York            22.00
3         London             8.70
4          Tokyo            13.50
5          Seoul            10.30
6  San francisco             8.75
7         Sydney             5.20
8          Dubai             3.38

In [5]:
#Using index parameter to change the index
indexes = ['I', 'II', 'III', 'IV', 'V', 'VI', 'VII', 'VIII', 'IX']
cities = pd.DataFrame({'City_names': list(city_names), 'Population_in_M' : list(population_in_M)}, index=indexes)
cities

City_names  Population_in_M
I         New Delhi            18.70
II           Mumbai            23.60
III        New York            22.00
IV           London             8.70
V             Tokyo            13.50
VI            Seoul            10.30
VII   San francisco             8.75
VIII         Sydney             5.20
IX            Dubai             3.38

## Reading from csv

In [6]:
california_housing_dataframe = pd.read_csv("https://download.mlcc.google.com/mledu-datasets/california_housing_train.csv", sep=",")
#To see first few sample from the dataframe
california_housing_dataframe.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -114.31     34.19                15.0       5612.0          1283.0   
1    -114.47     34.40                19.0       7650.0          1901.0   
2    -114.56     33.69                17.0        720.0           174.0   
3    -114.57     33.64                14.0       1501.0           337.0   
4    -114.57     33.57                20.0       1454.0           326.0   

   population  households  median_income  median_house_value  
0      1015.0       472.0         1.4936             66900.0  
1      1129.0       463.0         1.8200             80100.0  
2       333.0       117.0         1.6509             85700.0  
3       515.0       226.0         3.1917             73400.0  
4       624.0       262.0         1.9250             65500.0

In [7]:
california_housing_dataframe.head(10)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -114.31     34.19                15.0       5612.0          1283.0   
1    -114.47     34.40                19.0       7650.0          1901.0   
2    -114.56     33.69                17.0        720.0           174.0   
3    -114.57     33.64                14.0       1501.0           337.0   
4    -114.57     33.57                20.0       1454.0           326.0   
5    -114.58     33.63                29.0       1387.0           236.0   
6    -114.58     33.61                25.0       2907.0           680.0   
7    -114.59     34.83                41.0        812.0           168.0   
8    -114.59     33.61                34.0       4789.0          1175.0   
9    -114.60     34.83                46.0       1497.0           309.0   

   population  households  median_income  median_house_value  
0      1015.0       472.0         1.4936             66900.0  
1      1129.0       463.0         1.8200             80100.0  
2       333.0       117.0         1.6509             85700.0  
3       515.0       226.0         3.1917             73400.0  
4       624.0       262.0         1.9250             65500.0  
5       671.0       239.0         3.3438             74000.0  
6      1841.0       633.0         2.6768             82400.0  
7       375.0       158.0         1.7083             48500.0  
8      3134.0      1056.0         2.1782             58400.0  
9       787.0       271.0         2.1908             48100.0

In [8]:
#Gves the overall shpae of the entire dataframe
california_housing_dataframe.shape

(17000, 9)

In [9]:
#Getting statistical description of the data
california_housing_dataframe.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  17000.000000  17000.000000        17000.000000  17000.000000   
mean    -119.562108     35.625225           28.589353   2643.664412   
std        2.005166      2.137340           12.586937   2179.947071   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.790000     33.930000           18.000000   1462.000000   
50%     -118.490000     34.250000           29.000000   2127.000000   
75%     -118.000000     37.720000           37.000000   3151.250000   
max     -114.310000     41.950000           52.000000  37937.000000   

       total_bedrooms    population    households  median_income  \
count    17000.000000  17000.000000  17000.000000   17000.000000   
mean       539.410824   1429.573941    501.221941       3.883578   
std        421.499452   1147.852959    384.520841       1.908157   
min          1.000000      3.000000      1.000000       0.499900   
25%        297.000000    790.000000    282.000000       2.566375   
50%        434.000000   1167.000000    409.000000       3.544600   
75%        648.250000   1721.000000    605.250000       4.767000   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        17000.000000  
mean        207300.912353  
std         115983.764387  
min          14999.000000  
25%         119400.000000  
50%         180400.000000  
75%         265000.000000  
max         500001.000000

## Indexes

In [10]:
cities

City_names  Population_in_M
I         New Delhi            18.70
II           Mumbai            23.60
III        New York            22.00
IV           London             8.70
V             Tokyo            13.50
VI            Seoul            10.30
VII   San francisco             8.75
VIII         Sydney             5.20
IX            Dubai             3.38

In [11]:
cities.index

Index(['I', 'II', 'III', 'IV', 'V', 'VI', 'VII', 'VIII', 'IX'], dtype='object')

In [12]:
california_housing_dataframe.index

RangeIndex(start=0, stop=17000, step=1)

In [13]:
cities['Initials'] = ['ND', 'M', 'NY', 'L', 'T', 'SE', 'SF', 'SY', 'D']
cities


City_names  Population_in_M Initials
I         New Delhi            18.70       ND
II           Mumbai            23.60        M
III        New York            22.00       NY
IV           London             8.70        L
V             Tokyo            13.50        T
VI            Seoul            10.30       SE
VII   San francisco             8.75       SF
VIII         Sydney             5.20       SY
IX            Dubai             3.38        D

In [14]:
cities.reindex(np.random.permutation(cities.index))

City_names  Population_in_M Initials
IV           London             8.70        L
I         New Delhi            18.70       ND
VIII         Sydney             5.20       SY
V             Tokyo            13.50        T
VII   San francisco             8.75       SF
III        New York            22.00       NY
II           Mumbai            23.60        M
VI            Seoul            10.30       SE
IX            Dubai             3.38        D

In [15]:
#Setting and changing indexes
cities.set_index("Initials")

City_names  Population_in_M
Initials                                
ND            New Delhi            18.70
M                Mumbai            23.60
NY             New York            22.00
L                London             8.70
T                 Tokyo            13.50
SE                Seoul            10.30
SF        San francisco             8.75
SY               Sydney             5.20
D                 Dubai             3.38

## Accessing data from Dataframe

In [16]:
#Accessing an entire column based on a variable/column name the python dictionary way
cities['Population_in_M']

I       18.70
II      23.60
III     22.00
IV       8.70
V       13.50
VI      10.30
VII      8.75
VIII     5.20
IX       3.38
Name: Population_in_M, dtype: float64

In [17]:
type(cities['Population_in_M'])

pandas.core.series.Series

In [18]:
#Naive way
cities.City_names

I           New Delhi
II             Mumbai
III          New York
IV             London
V               Tokyo
VI              Seoul
VII     San francisco
VIII           Sydney
IX              Dubai
Name: City_names, dtype: object

In [19]:
#Accessing particular cell from a column using index
cities['City_names'][2]

'New York'

In [20]:
type(cities['City_names'][2])

str

In [21]:
# Accessing a varuable for it's description
california_housing_dataframe.total_bedrooms.describe()

count    17000.000000
mean       539.410824
std        421.499452
min          1.000000
25%        297.000000
50%        434.000000
75%        648.250000
max       6445.000000
Name: total_bedrooms, dtype: float64

In [22]:
california_housing_dataframe.total_bedrooms.mean()

539.4108235294118

In [23]:
cities.City_names.describe()

count         9
unique        9
top       Dubai
freq          1
Name: City_names, dtype: object

In [24]:
cities.City_names.unique()

array(['New Delhi', 'Mumbai', 'New York', 'London', 'Tokyo', 'Seoul',
       'San francisco', 'Sydney', 'Dubai'], dtype=object)

In [25]:
cities.City_names.value_counts()

Dubai            1
New Delhi        1
Seoul            1
London           1
San francisco    1
Mumbai           1
Sydney           1
New York         1
Tokyo            1
Name: City_names, dtype: int64

### Indexing in Pandas

#### iloc operator: index-based selection

In [26]:
#Accessing a single row from dataframe
cities.iloc[1]

City_names         Mumbai
Population_in_M      23.6
Initials                M
Name: II, dtype: object

In [27]:
type(cities.iloc[1])

pandas.core.series.Series

In [28]:
cities.iloc[-1]

City_names         Dubai
Population_in_M     3.38
Initials               D
Name: IX, dtype: object

In [29]:
california_housing_dataframe.iloc[25:29]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
25    -115.32     32.82                34.0        591.0           139.0   
26    -115.37     32.82                30.0       1602.0           322.0   
27    -115.37     32.82                14.0       1276.0           270.0   
28    -115.37     32.81                32.0        741.0           191.0   

    population  households  median_income  median_house_value  
25       327.0        89.0         3.6528            100000.0  
26      1130.0       335.0         3.5735             71100.0  
27       867.0       261.0         1.9375             80900.0  
28       623.0       169.0         1.7604             68600.0

In [30]:
type(california_housing_dataframe.iloc[0:3])

pandas.core.frame.DataFrame

In [31]:
# Retrieving an entirecolumn using iloc
cities.iloc[:,0]

I           New Delhi
II             Mumbai
III          New York
IV             London
V               Tokyo
VI              Seoul
VII     San francisco
VIII           Sydney
IX              Dubai
Name: City_names, dtype: object

In [32]:
# Fetching few elements from a column
# Fetching first 10 entries from total_rooms
california_housing_dataframe.iloc[:10, 3]

0    5612.0
1    7650.0
2     720.0
3    1501.0
4    1454.0
5    1387.0
6    2907.0
7     812.0
8    4789.0
9    1497.0
Name: total_rooms, dtype: float64

In [33]:
#Accessing a list of entries
cities.iloc[[2, 4,6],0]

III         New York
V              Tokyo
VII    San francisco
Name: City_names, dtype: object

#### loc operator: label-based selection.

In [34]:
cities.loc['II', 'City_names']

'Mumbai'

In [35]:
california_housing_dataframe.loc[2:5, 'population']

2    333.0
3    515.0
4    624.0
5    671.0
Name: population, dtype: float64

In [36]:
california_housing_dataframe.loc[2:5, ['latitude', 'longitude']]

latitude  longitude
2     33.69    -114.56
3     33.64    -114.57
4     33.57    -114.57
5     33.63    -114.58

*iloc uses the Python stdlib indexing scheme, where the first element of the range is included and the last one excluded. So 0:10 will select entries 0,...,9. loc, meanwhile, indexes inclusively. So 0:10 will select entries 0,...,10.*

**Use *loc* when index are alpha-numeric or just words, and use *iloc* when indexes are only numeric**

### Conditional Selection

In [37]:
california_housing_dataframe.total_bedrooms > 500

0         True
1         True
2        False
3        False
4        False
         ...  
16995    False
16996     True
16997     True
16998     True
16999    False
Name: total_bedrooms, Length: 17000, dtype: bool

In [38]:
california_housing_dataframe.loc[california_housing_dataframe.total_bedrooms > 1000]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -114.31     34.19                15.0       5612.0          1283.0   
1        -114.47     34.40                19.0       7650.0          1901.0   
8        -114.59     33.61                34.0       4789.0          1175.0   
115      -115.85     34.20                34.0       3868.0          1257.0   
133      -116.06     34.15                15.0      10377.0          2331.0   
...          ...       ...                 ...          ...             ...   
16824    -123.22     39.28                16.0       5569.0          1106.0   
16838    -123.34     39.10                24.0       5372.0          1051.0   
16854    -123.49     38.70                 9.0       5409.0          1019.0   
16859    -123.59     38.80                17.0       5202.0          1037.0   
16986    -124.19     40.73                21.0       5694.0          1056.0   

       population  households  median_income  median_house_value  
0          1015.0       472.0         1.4936             66900.0  
1          1129.0       463.0         1.8200             80100.0  
8          3134.0      1056.0         2.1782             58400.0  
115         890.0       423.0         1.3571             41000.0  
133        4507.0      1807.0         2.2466             66800.0  
...           ...         ...            ...                 ...  
16824      3148.0      1088.0         3.1455            142900.0  
16838      3002.0       992.0         3.0652            131100.0  
16854       594.0       327.0         3.3125            295400.0  
16859      1742.0       803.0         3.1201            176100.0  
16986      2907.0       972.0         3.5363             90100.0  

[1576 rows x 9 columns]

Only 1576 out of 17,000 (somewhere around 9%) areas/blocks/locality have number of bedrooms above 1000

In [39]:
# Logical connectors & | ! can be used to combine multiple conditions
california_housing_dataframe.loc[(california_housing_dataframe.total_bedrooms > 1000) & (california_housing_dataframe.total_rooms > 5000)]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -114.31     34.19                15.0       5612.0          1283.0   
1        -114.47     34.40                19.0       7650.0          1901.0   
133      -116.06     34.15                15.0      10377.0          2331.0   
135      -116.09     34.15                13.0       9444.0          1997.0   
167      -116.24     33.72                25.0       5236.0          1039.0   
...          ...       ...                 ...          ...             ...   
16824    -123.22     39.28                16.0       5569.0          1106.0   
16838    -123.34     39.10                24.0       5372.0          1051.0   
16854    -123.49     38.70                 9.0       5409.0          1019.0   
16859    -123.59     38.80                17.0       5202.0          1037.0   
16986    -124.19     40.73                21.0       5694.0          1056.0   

       population  households  median_income  median_house_value  
0          1015.0       472.0         1.4936             66900.0  
1          1129.0       463.0         1.8200             80100.0  
133        4507.0      1807.0         2.2466             66800.0  
135        4166.0      1482.0         2.6111             65600.0  
167        2725.0       935.0         3.7750             93400.0  
...           ...         ...            ...                 ...  
16824      3148.0      1088.0         3.1455            142900.0  
16838      3002.0       992.0         3.0652            131100.0  
16854       594.0       327.0         3.3125            295400.0  
16859      1742.0       803.0         3.1201            176100.0  
16986      2907.0       972.0         3.5363             90100.0  

[1169 rows x 9 columns]

Some other usefull functions to explore: *isin()*, *isnull()*, *notnull()*, 

## Manipulating Data

In [40]:
cities['Population_in_M'] *= 2 

In [41]:
cities

City_names  Population_in_M Initials
I         New Delhi            37.40       ND
II           Mumbai            47.20        M
III        New York            44.00       NY
IV           London            17.40        L
V             Tokyo            27.00        T
VI            Seoul            20.60       SE
VII   San francisco            17.50       SF
VIII         Sydney            10.40       SY
IX            Dubai             6.76        D

In [42]:
cities['Area_Square_km'] = list(pd.Series([1484, 603.4, 783.8, 1572, 2188, 605.2, 121.4, 1687, 4114]))
cities['population_density'] = cities['Population_in_M'] / cities['Area_Square_km']

In [43]:
cities

City_names  Population_in_M Initials  Area_Square_km  \
I         New Delhi            37.40       ND          1484.0   
II           Mumbai            47.20        M           603.4   
III        New York            44.00       NY           783.8   
IV           London            17.40        L          1572.0   
V             Tokyo            27.00        T          2188.0   
VI            Seoul            20.60       SE           605.2   
VII   San francisco            17.50       SF           121.4   
VIII         Sydney            10.40       SY          1687.0   
IX            Dubai             6.76        D          4114.0   

      population_density  
I               0.025202  
II              0.078223  
III             0.056137  
IV              0.011069  
V               0.012340  
VI              0.034038  
VII             0.144152  
VIII            0.006165  
IX              0.001643

#### Using map()
map takes a single value from a series and returns the manipulates the value. In the end returns a manipulated Series.

In [44]:
population_mean = cities.Population_in_M.mean()
cities['Away_from_mean_population'] = cities.Population_in_M.map(lambda p : p - population_mean)

In [45]:
cities

City_names  Population_in_M Initials  Area_Square_km  \
I         New Delhi            37.40       ND          1484.0   
II           Mumbai            47.20        M           603.4   
III        New York            44.00       NY           783.8   
IV           London            17.40        L          1572.0   
V             Tokyo            27.00        T          2188.0   
VI            Seoul            20.60       SE           605.2   
VII   San francisco            17.50       SF           121.4   
VIII         Sydney            10.40       SY          1687.0   
IX            Dubai             6.76        D          4114.0   

      population_density  Away_from_mean_population  
I               0.025202                  12.037778  
II              0.078223                  21.837778  
III             0.056137                  18.637778  
IV              0.011069                  -7.962222  
V               0.012340                   1.637778  
VI              0.034038                  -4.762222  
VII             0.144152                  -7.862222  
VIII            0.006165                 -14.962222  
IX              0.001643                 -18.602222

apply() is the equivalent method if we want to transform a whole DataFrame by calling a custom method on each row.

In [46]:
cities['Country'] = ['India', 'India', 'USA', 'UK', 'Japan', 'South Korea', 'USA', 'Australia', 'UAE']
cities

City_names  Population_in_M Initials  Area_Square_km  \
I         New Delhi            37.40       ND          1484.0   
II           Mumbai            47.20        M           603.4   
III        New York            44.00       NY           783.8   
IV           London            17.40        L          1572.0   
V             Tokyo            27.00        T          2188.0   
VI            Seoul            20.60       SE           605.2   
VII   San francisco            17.50       SF           121.4   
VIII         Sydney            10.40       SY          1687.0   
IX            Dubai             6.76        D          4114.0   

      population_density  Away_from_mean_population      Country  
I               0.025202                  12.037778        India  
II              0.078223                  21.837778        India  
III             0.056137                  18.637778          USA  
IV              0.011069                  -7.962222           UK  
V               0.012340                   1.637778        Japan  
VI              0.034038                  -4.762222  South Korea  
VII             0.144152                  -7.862222          USA  
VIII            0.006165                 -14.962222    Australia  
IX              0.001643                 -18.602222          UAE

In [47]:
cities.Country+'--'+cities.City_names

I         India--New Delhi
II           India--Mumbai
III          USA--New York
IV              UK--London
V             Japan--Tokyo
VI      South Korea--Seoul
VII     USA--San francisco
VIII     Australia--Sydney
IX              UAE--Dubai
dtype: object

In [48]:
cities.population_density.sum()

0.36896892424153005

Explore other manipulations as and when needed

In [49]:
california_housing_dataframe.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value'],
      dtype='object')

In [50]:
california_housing_dataframe.total_bedrooms.count() ,len(california_housing_dataframe.total_bedrooms.unique())

(17000, 1848)

In [51]:
california_housing_dataframe.total_bedrooms.value_counts()

280.0     48
309.0     44
394.0     43
331.0     43
345.0     43
          ..
961.0      1
3358.0     1
2537.0     1
2366.0     1
2062.0     1
Name: total_bedrooms, Length: 1848, dtype: int64

## Grouping and Analyzing

In [52]:
cities.groupby('Country')

In [53]:
cities.groupby('Country').City_names

In [54]:
#Equivalent of count_values()
cities.groupby('Country').Country.count()

Country
Australia      1
India          2
Japan          1
South Korea    1
UAE            1
UK             1
USA            2
Name: Country, dtype: int64

groupby() created a slice of data from cities based on *Country*. So cities with same country got grouped in 1 entry.
In the above example, for each of the group, in the Country column, we see how many times each entry appears.

In [55]:
#Any summary functions can be used
cities.groupby('Country').Population_in_M.mean()

Country
Australia      10.40
India          42.30
Japan          27.00
South Korea    20.60
UAE             6.76
UK             17.40
USA            30.75
Name: Population_in_M, dtype: float64

In [56]:
cities.groupby('Country').apply(lambda df: df.loc[df.Population_in_M.idxmax()])

City_names  Population_in_M Initials  Area_Square_km  \
Country                                                            
Australia       Sydney            10.40       SY          1687.0   
India           Mumbai            47.20        M           603.4   
Japan            Tokyo            27.00        T          2188.0   
South Korea      Seoul            20.60       SE           605.2   
UAE              Dubai             6.76        D          4114.0   
UK              London            17.40        L          1572.0   
USA           New York            44.00       NY           783.8   

             population_density  Away_from_mean_population      Country  
Country                                                                  
Australia              0.006165                 -14.962222    Australia  
India                  0.078223                  21.837778        India  
Japan                  0.012340                   1.637778        Japan  
South Korea            0.034038                  -4.762222  South Korea  
UAE                    0.001643                 -18.602222          UAE  
UK                     0.011069                  -7.962222           UK  
USA                    0.056137                  18.637778          USA

Think of each group we generate as being a slice of our DataFrame containing only data with values that match.

In [57]:
california_housing_dataframe.groupby(['longitude', 'latitude']).apply(lambda df : df.loc[df.total_rooms.idxmax()])

longitude  latitude  housing_median_age  total_rooms  \
longitude latitude                                                         
-124.35   40.54       -124.35     40.54                52.0       1820.0   
-124.30   41.80       -124.30     41.80                19.0       2672.0   
          41.84       -124.30     41.84                17.0       2677.0   
-124.27   40.69       -124.27     40.69                36.0       2349.0   
-124.26   40.58       -124.26     40.58                52.0       2217.0   
...                       ...       ...                 ...          ...   
-114.57   33.57       -114.57     33.57                20.0       1454.0   
          33.64       -114.57     33.64                14.0       1501.0   
-114.56   33.69       -114.56     33.69                17.0        720.0   
-114.47   34.40       -114.47     34.40                19.0       7650.0   
-114.31   34.19       -114.31     34.19                15.0       5612.0   

                    total_bedrooms  population  households  median_income  \
longitude latitude                                                          
-124.35   40.54              300.0       806.0       270.0         3.0147   
-124.30   41.80              552.0      1298.0       478.0         1.9797   
          41.84              531.0      1244.0       456.0         3.0313   
-124.27   40.69              528.0      1194.0       465.0         2.5179   
-124.26   40.58              394.0       907.0       369.0         2.3571   
...                            ...         ...         ...            ...   
-114.57   33.57              326.0       624.0       262.0         1.9250   
          33.64              337.0       515.0       226.0         3.1917   
-114.56   33.69              174.0       333.0       117.0         1.6509   
-114.47   34.40             1901.0      1129.0       463.0         1.8200   
-114.31   34.19             1283.0      1015.0       472.0         1.4936   

                    median_house_value  
longitude latitude                      
-124.35   40.54                94600.0  
-124.30   41.80                85800.0  
          41.84               103600.0  
-124.27   40.69                79000.0  
-124.26   40.58               111400.0  
...                                ...  
-114.57   33.57                65500.0  
          33.64                73400.0  
-114.56   33.69                85700.0  
-114.47   34.40                80100.0  
-114.31   34.19                66900.0  

[11054 rows x 9 columns]

In [58]:
#Generating a simple summary statistics using agg() function
groupDF = cities.groupby('Country').Population_in_M.agg([len, 'min', 'max', 'mean', 'sum', 'var', 'std'])
groupDF

len    min    max   mean    sum      var        std
Country                                                         
Australia    1.0  10.40  10.40  10.40  10.40      NaN        NaN
India        2.0  37.40  47.20  42.30  84.60   48.020   6.929646
Japan        1.0  27.00  27.00  27.00  27.00      NaN        NaN
South Korea  1.0  20.60  20.60  20.60  20.60      NaN        NaN
UAE          1.0   6.76   6.76   6.76   6.76      NaN        NaN
UK           1.0  17.40  17.40  17.40  17.40      NaN        NaN
USA          2.0  17.50  44.00  30.75  61.50  351.125  18.738330

Check the link for list of aggregate functions <bt>
    https://cmdlinetips.com/2019/10/pandas-groupby-13-functions-to-aggregate/

### Sorting

In [59]:
cities.sort_values(by='Initials')

City_names  Population_in_M Initials  Area_Square_km  \
IX            Dubai             6.76        D          4114.0   
IV           London            17.40        L          1572.0   
II           Mumbai            47.20        M           603.4   
I         New Delhi            37.40       ND          1484.0   
III        New York            44.00       NY           783.8   
VI            Seoul            20.60       SE           605.2   
VII   San francisco            17.50       SF           121.4   
VIII         Sydney            10.40       SY          1687.0   
V             Tokyo            27.00        T          2188.0   

      population_density  Away_from_mean_population      Country  
IX              0.001643                 -18.602222          UAE  
IV              0.011069                  -7.962222           UK  
II              0.078223                  21.837778        India  
I               0.025202                  12.037778        India  
III             0.056137                  18.637778          USA  
VI              0.034038                  -4.762222  South Korea  
VII             0.144152                  -7.862222          USA  
VIII            0.006165                 -14.962222    Australia  
V               0.012340                   1.637778        Japan

In [60]:
cities.sort_index()

City_names  Population_in_M Initials  Area_Square_km  \
I         New Delhi            37.40       ND          1484.0   
II           Mumbai            47.20        M           603.4   
III        New York            44.00       NY           783.8   
IV           London            17.40        L          1572.0   
IX            Dubai             6.76        D          4114.0   
V             Tokyo            27.00        T          2188.0   
VI            Seoul            20.60       SE           605.2   
VII   San francisco            17.50       SF           121.4   
VIII         Sydney            10.40       SY          1687.0   

      population_density  Away_from_mean_population      Country  
I               0.025202                  12.037778        India  
II              0.078223                  21.837778        India  
III             0.056137                  18.637778          USA  
IV              0.011069                  -7.962222           UK  
IX              0.001643                 -18.602222          UAE  
V               0.012340                   1.637778        Japan  
VI              0.034038                  -4.762222  South Korea  
VII             0.144152                  -7.862222          USA  
VIII            0.006165                 -14.962222    Australia

### Data Types of elements inside dataframe

In [61]:
#using dtype property to get the data type of an element of individual column
cities.Population_in_M.dtype

dtype('float64')

In [62]:
#using dtypes with dataframe object gives data types for all the columns in the dataframe
cities.dtypes

City_names                    object
Population_in_M              float64
Initials                      object
Area_Square_km               float64
population_density           float64
Away_from_mean_population    float64
Country                       object
dtype: object

In [70]:
cities.index.dtype

dtype('O')

#### Missing data
Missing values: **NaN** <br>
NaN: *float64*

In [77]:
cities["Indian_Restaurant"] = [1234, 2345, 123, 234, 12, None, 254, 345, 1991]
cities

City_names  Population_in_M Initials  Area_Square_km  \
I         New Delhi            37.40       ND          1484.0   
II           Mumbai            47.20        M           603.4   
III        New York            44.00       NY           783.8   
IV           London            17.40        L          1572.0   
V             Tokyo            27.00        T          2188.0   
VI            Seoul            20.60       SE           605.2   
VII   San francisco            17.50       SF           121.4   
VIII         Sydney            10.40       SY          1687.0   
IX            Dubai             6.76        D          4114.0   

      population_density  Away_from_mean_population      Country  \
I               0.025202                  12.037778        India   
II              0.078223                  21.837778        India   
III             0.056137                  18.637778          USA   
IV              0.011069                  -7.962222           UK   
V               0.012340                   1.637778        Japan   
VI              0.034038                  -4.762222  South Korea   
VII             0.144152                  -7.862222          USA   
VIII            0.006165                 -14.962222    Australia   
IX              0.001643                 -18.602222          UAE   

      Indian_Restaurant  
I                1234.0  
II               2345.0  
III               123.0  
IV                234.0  
V                  12.0  
VI                  NaN  
VII               254.0  
VIII              345.0  
IX               1991.0

In [78]:
#pd.isnull(): Checks whether the value is NaN or not
#Complementary function is pd.notnull()
cities.isnull()

City_names  Population_in_M  Initials  Area_Square_km  \
I          False            False     False           False   
II         False            False     False           False   
III        False            False     False           False   
IV         False            False     False           False   
V          False            False     False           False   
VI         False            False     False           False   
VII        False            False     False           False   
VIII       False            False     False           False   
IX         False            False     False           False   

      population_density  Away_from_mean_population  Country  \
I                  False                      False    False   
II                 False                      False    False   
III                False                      False    False   
IV                 False                      False    False   
V                  False                      False    False   
VI                 False                      False    False   
VII                False                      False    False   
VIII               False                      False    False   
IX                 False                      False    False   

      Indian_Restaurant  
I                 False  
II                False  
III               False  
IV                False  
V                 False  
VI                 True  
VII               False  
VIII              False  
IX                False

In [81]:
#Finding Nan entries specific to a column
cities[pd.isnull(cities.Indian_Restaurant)]

City_names  Population_in_M Initials  Area_Square_km  population_density  \
VI      Seoul             20.6       SE           605.2            0.034038   

    Away_from_mean_population      Country  Indian_Restaurant  
VI                  -4.762222  South Korea                NaN

In [83]:
#Changing type for column Indian_restaurant
#cities.Indian_Restaurant.astype('int64')

In [88]:
#Changing NaN with desired type
cities.Indian_Restaurant.fillna(0)

I       1234.0
II      2345.0
III      123.0
IV       234.0
V         12.0
VI         0.0
VII      254.0
VIII     345.0
IX      1991.0
Name: Indian_Restaurant, dtype: float64

In [90]:
cities

City_names  Population_in_M Initials  Area_Square_km  \
I         New Delhi            37.40       ND          1484.0   
II           Mumbai            47.20        M           603.4   
III        New York            44.00       NY           783.8   
IV           London            17.40        L          1572.0   
V             Tokyo            27.00        T          2188.0   
VI            Seoul            20.60       SE           605.2   
VII   San francisco            17.50       SF           121.4   
VIII         Sydney            10.40       SY          1687.0   
IX            Dubai             6.76        D          4114.0   

      population_density  Away_from_mean_population      Country  \
I               0.025202                  12.037778        India   
II              0.078223                  21.837778        India   
III             0.056137                  18.637778          USA   
IV              0.011069                  -7.962222           UK   
V               0.012340                   1.637778        Japan   
VI              0.034038                  -4.762222  South Korea   
VII             0.144152                  -7.862222          USA   
VIII            0.006165                 -14.962222    Australia   
IX              0.001643                 -18.602222          UAE   

      Indian_Restaurant  
I                1234.0  
II               2345.0  
III               123.0  
IV                234.0  
V                  12.0  
VI                  NaN  
VII               254.0  
VIII              345.0  
IX               1991.0

In [91]:
#Modifies the dataframe as well
cities.Indian_Restaurant.fillna(0, inplace=True)

In [92]:
cities

City_names  Population_in_M Initials  Area_Square_km  \
I         New Delhi            37.40       ND          1484.0   
II           Mumbai            47.20        M           603.4   
III        New York            44.00       NY           783.8   
IV           London            17.40        L          1572.0   
V             Tokyo            27.00        T          2188.0   
VI            Seoul            20.60       SE           605.2   
VII   San francisco            17.50       SF           121.4   
VIII         Sydney            10.40       SY          1687.0   
IX            Dubai             6.76        D          4114.0   

      population_density  Away_from_mean_population      Country  \
I               0.025202                  12.037778        India   
II              0.078223                  21.837778        India   
III             0.056137                  18.637778          USA   
IV              0.011069                  -7.962222           UK   
V               0.012340                   1.637778        Japan   
VI              0.034038                  -4.762222  South Korea   
VII             0.144152                  -7.862222          USA   
VIII            0.006165                 -14.962222    Australia   
IX              0.001643                 -18.602222          UAE   

      Indian_Restaurant  
I                1234.0  
II               2345.0  
III               123.0  
IV                234.0  
V                  12.0  
VI                  0.0  
VII               254.0  
VIII              345.0  
IX               1991.0

In [94]:
cities.Indian_Restaurant = cities.Indian_Restaurant.astype('int64')

In [95]:
cities

City_names  Population_in_M Initials  Area_Square_km  \
I         New Delhi            37.40       ND          1484.0   
II           Mumbai            47.20        M           603.4   
III        New York            44.00       NY           783.8   
IV           London            17.40        L          1572.0   
V             Tokyo            27.00        T          2188.0   
VI            Seoul            20.60       SE           605.2   
VII   San francisco            17.50       SF           121.4   
VIII         Sydney            10.40       SY          1687.0   
IX            Dubai             6.76        D          4114.0   

      population_density  Away_from_mean_population      Country  \
I               0.025202                  12.037778        India   
II              0.078223                  21.837778        India   
III             0.056137                  18.637778          USA   
IV              0.011069                  -7.962222           UK   
V               0.012340                   1.637778        Japan   
VI              0.034038                  -4.762222  South Korea   
VII             0.144152                  -7.862222          USA   
VIII            0.006165                 -14.962222    Australia   
IX              0.001643                 -18.602222          UAE   

      Indian_Restaurant  
I                  1234  
II                 2345  
III                 123  
IV                  234  
V                    12  
VI                    0  
VII                 254  
VIII                345  
IX                 1991

### Replacing an entry

In [97]:
cities.Country = cities.Country.replace('USA', 'America')
cities

City_names  Population_in_M Initials  Area_Square_km  \
I         New Delhi            37.40       ND          1484.0   
II           Mumbai            47.20        M           603.4   
III        New York            44.00       NY           783.8   
IV           London            17.40        L          1572.0   
V             Tokyo            27.00        T          2188.0   
VI            Seoul            20.60       SE           605.2   
VII   San francisco            17.50       SF           121.4   
VIII         Sydney            10.40       SY          1687.0   
IX            Dubai             6.76        D          4114.0   

      population_density  Away_from_mean_population      Country  \
I               0.025202                  12.037778        India   
II              0.078223                  21.837778        India   
III             0.056137                  18.637778      America   
IV              0.011069                  -7.962222           UK   
V               0.012340                   1.637778        Japan   
VI              0.034038                  -4.762222  South Korea   
VII             0.144152                  -7.862222      America   
VIII            0.006165                 -14.962222    Australia   
IX              0.001643                 -18.602222          UAE   

      Indian_Restaurant  
I                  1234  
II                 2345  
III                 123  
IV                  234  
V                    12  
VI                    0  
VII                 254  
VIII                345  
IX                 1991

### Renaming

In [99]:
#Renaming a column name in a dataframe
cities = cities.rename(columns={'Population_in_M': 'Population', 'Area_Square_km': 'Area'})
cities

City_names  Population Initials    Area  population_density  \
I         New Delhi       37.40       ND  1484.0            0.025202   
II           Mumbai       47.20        M   603.4            0.078223   
III        New York       44.00       NY   783.8            0.056137   
IV           London       17.40        L  1572.0            0.011069   
V             Tokyo       27.00        T  2188.0            0.012340   
VI            Seoul       20.60       SE   605.2            0.034038   
VII   San francisco       17.50       SF   121.4            0.144152   
VIII         Sydney       10.40       SY  1687.0            0.006165   
IX            Dubai        6.76        D  4114.0            0.001643   

      Away_from_mean_population      Country  Indian_Restaurant  
I                     12.037778        India               1234  
II                    21.837778        India               2345  
III                   18.637778      America                123  
IV                    -7.962222           UK                234  
V                      1.637778        Japan                 12  
VI                    -4.762222  South Korea                  0  
VII                   -7.862222      America                254  
VIII                 -14.962222    Australia                345  
IX                   -18.602222          UAE               1991

In [100]:
#Renaming vaues at index
cities.rename(index={'I':'First', 'IX':'Last'})

City_names  Population Initials    Area  population_density  \
First      New Delhi       37.40       ND  1484.0            0.025202   
II            Mumbai       47.20        M   603.4            0.078223   
III         New York       44.00       NY   783.8            0.056137   
IV            London       17.40        L  1572.0            0.011069   
V              Tokyo       27.00        T  2188.0            0.012340   
VI             Seoul       20.60       SE   605.2            0.034038   
VII    San francisco       17.50       SF   121.4            0.144152   
VIII          Sydney       10.40       SY  1687.0            0.006165   
Last           Dubai        6.76        D  4114.0            0.001643   

       Away_from_mean_population      Country  Indian_Restaurant  
First                  12.037778        India               1234  
II                     21.837778        India               2345  
III                    18.637778      America                123  
IV                     -7.962222           UK                234  
V                       1.637778        Japan                 12  
VI                     -4.762222  South Korea                  0  
VII                    -7.862222      America                254  
VIII                  -14.962222    Australia                345  
Last                  -18.602222          UAE               1991

explore rename_axis() method.

### Merging Dataframes or Series

#### concat()
Given a list of elements, this function will smush those elements together along an axis.
<br>
This is useful when we have data in different DataFrame or Series objects but having the same fields (columns). 

In [106]:
df1 = cities.iloc[[1,3,5,7]]
df1

City_names  Population Initials    Area  population_density  \
II       Mumbai        47.2        M   603.4            0.078223   
IV       London        17.4        L  1572.0            0.011069   
VI        Seoul        20.6       SE   605.2            0.034038   
VIII     Sydney        10.4       SY  1687.0            0.006165   

      Away_from_mean_population      Country  Indian_Restaurant  
II                    21.837778        India               2345  
IV                    -7.962222           UK                234  
VI                    -4.762222  South Korea                  0  
VIII                 -14.962222    Australia                345

In [107]:
df2 = cities.iloc[[0,2,4,6,8]]
df2

City_names  Population Initials    Area  population_density  \
I        New Delhi       37.40       ND  1484.0            0.025202   
III       New York       44.00       NY   783.8            0.056137   
V            Tokyo       27.00        T  2188.0            0.012340   
VII  San francisco       17.50       SF   121.4            0.144152   
IX           Dubai        6.76        D  4114.0            0.001643   

     Away_from_mean_population  Country  Indian_Restaurant  
I                    12.037778    India               1234  
III                  18.637778  America                123  
V                     1.637778    Japan                 12  
VII                  -7.862222  America                254  
IX                  -18.602222      UAE               1991

In [109]:
pd.concat([df1, df2])

City_names  Population Initials    Area  population_density  \
II           Mumbai       47.20        M   603.4            0.078223   
IV           London       17.40        L  1572.0            0.011069   
VI            Seoul       20.60       SE   605.2            0.034038   
VIII         Sydney       10.40       SY  1687.0            0.006165   
I         New Delhi       37.40       ND  1484.0            0.025202   
III        New York       44.00       NY   783.8            0.056137   
V             Tokyo       27.00        T  2188.0            0.012340   
VII   San francisco       17.50       SF   121.4            0.144152   
IX            Dubai        6.76        D  4114.0            0.001643   

      Away_from_mean_population      Country  Indian_Restaurant  
II                    21.837778        India               2345  
IV                    -7.962222           UK                234  
VI                    -4.762222  South Korea                  0  
VIII                 -14.962222    Australia                345  
I                     12.037778        India               1234  
III                   18.637778      America                123  
V                      1.637778        Japan                 12  
VII                   -7.862222      America                254  
IX                   -18.602222          UAE               1991

#### join()
join() lets you combine different DataFrame objects which have an index in common.

In [113]:
df3 = cities.loc[:,['City_names', 'Initials']]
df3

City_names Initials
I         New Delhi       ND
II           Mumbai        M
III        New York       NY
IV           London        L
V             Tokyo        T
VI            Seoul       SE
VII   San francisco       SF
VIII         Sydney       SY
IX            Dubai        D

In [114]:
df4 = cities.loc[:,['City_names', 'Country']]
df4

City_names      Country
I         New Delhi        India
II           Mumbai        India
III        New York      America
IV           London           UK
V             Tokyo        Japan
VI            Seoul  South Korea
VII   San francisco      America
VIII         Sydney    Australia
IX            Dubai          UAE

In [116]:
df3.join(df4, lsuffix='_df3', rsuffix='_df4')

City_names_df3 Initials City_names_df4      Country
I         New Delhi       ND      New Delhi        India
II           Mumbai        M         Mumbai        India
III        New York       NY       New York      America
IV           London        L         London           UK
V             Tokyo        T          Tokyo        Japan
VI            Seoul       SE          Seoul  South Korea
VII   San francisco       SF  San francisco      America
VIII         Sydney       SY         Sydney    Australia
IX            Dubai        D          Dubai          UAE

In [118]:
df3.set_index('City_names').join(df4.set_index('City_names'))

Initials      Country
City_names                         
New Delhi           ND        India
Mumbai               M        India
New York            NY      America
London               L           UK
Tokyo                T        Japan
Seoul               SE  South Korea
San francisco       SF      America
Sydney              SY    Australia
Dubai                D          UAE

Explore merge(), most of th